### DISCUSSION TITLE
Expose a `ctx.response` object inside the `handler` callback in server actions

### DISCUSSION THREAD

**Discussion by NicoToff** [withastro/roadmap#1009] [Proposal]
### Body

# Summary

As a developer, when using server actions, it would be great to have a little more control over the `Response` that is returned to the client.

# Background & Motivation

Right now, server actions return JSON with a 200 status on success, or can throw errors with various 5xx statuses thanks to the `ActionError`.

However, for example, it is currently not possible to:
- Return a 201 code after a successful POST request
- Set specific HTTP headers on the response

# Goals

The goal of this request is to gain finer control over the `Response` object, so that server actions can cover more use cases.

The fields that should be available for setting on the `Response` are:
- `headers`
- `status`
- `statusText`

All fields of the response shouldn't be accessible to avoid some misuse (like trying to set `Response.body` manually).

# Example

Adding arbitrary HTTP headers:
```typescript
import { defineAction } from "astro:actions";

export const server = {
    getData: defineAction({
        handler: async (_, ctx) => {
            const data = { /* Getting some data... */ };
            ctx.response.headers = {
                'Cache-Control': 'public, no-cache',
                ETag: '' /* Generate ETag... */
            };
            return data;
        }
    })
};
```

Modifying a status code manually:
```typescript
import { defineAction } from 'astro:actions';

export const server = {
    createProfile: defineAction({
        input: { /* Some payload */ },
        handler: async (input, ctx) => {
            // Handing input...
            ctx.response.status = 201;
            ctx.response.statusText = "Created";
            return { success: true };
        }
    })
};
```

**Comment by Fryuni** [comment]
> All fields of the response shouldn't be accessible to avoid some misuse (like trying to set Response.body manually).

It might be simpler to expose the `ResponseInit` object, which will be passed to the `Response` constructor.

**Reply by NicoToff** [reply]
> > All fields of the response shouldn't be accessible to avoid some misuse (like trying to set Response.body manually).
> 
> It might be simpler to expose the `ResponseInit` object, which will be passed to the `Response` constructor.

Good point, this is indeed the data that is meant to be exposed (aka, the [`options` from the `Response` constructor](https://developer.mozilla.org/en-US/docs/Web/API/Response/Response).

In this case, maybe the instantiation of the `Response` is already done under the hood, so we only modify specific fields from it.

Anyway, the API and typing to modify those three fields can be debated, but I find that assigning to the fields directly makes it quite clear and straightforward to consumers. While exposing a `ctx.response.init` or `ctx.responseInit` object would be more cryptic.
